scipy/optimize/tests/test__numdiff.py

```diff
@@ -124,18 +124,21 @@ def jac_scalar_scalar(self, x):
         return np.cosh(x)
 
     def fun_scalar_vector(self, x):
-        return np.array([x[0]**2, np.tan(x[0]), np.exp(x[0])])
```

Comment:
Please don't make changes to the existing tests at the same time as changing the code, if the aim of the changes is to keep existing code working.

I think there are two things mixed in here that are separate: (i) cleaning up the dimension handing in numdiff, and (ii) dealing with possible(?) future numpy changes where it will start emitting warnings.

Reply:
I'm only looking at (i) here, (ii) was just the catalyst for spotting that the dimension handling was a little clumsy.

> Here it's a bit less important than elsewhere because IIRC _numdiff.py is not public API

That was my hope too. I can't cleanup `_numdiff.py` without changing the tests, because the tests are testing for the design flaws I'm trying to remove - and I don't think there's value to an intermediate commit that does not pass CI.

Reply:
Right, however, numdiff is called on user-provided functions in the optimizers, and for that looking at the tests, it is not clear whether it behaves the same now as before. There seem to be some behavior changes in this PR, so it's not just internal cleanup.